thermal-control-design.md

```diff
@@ -0,0 +1,272 @@
+
+
+  
+## 1. Overview
+
+The purpose of Thermal Control is to keep the switch at a proper temperature by using cooling devices, e.g., fan.
+Thermal control daemon need to monitor the temperature of devices (CPU, ASIC, optical modules, etc) and the running status of fan. It store temperature values fetched from sensors and thermal device running status to the DB, to make these data available to CLI and SNMP or other apps which interested. 
+
+Thermal control also enforce some environment related polices to help the thermal control algorithm to adjust the switch temperature.     
+
+## 2. Thermal device monitoring
+
+Thermal monitoring function will retrieve the switch device temperatures via platform APIs, platform APIs not only provide the value of the temperature, but also provide the threshold value. The thermal object status can be deduced by comparing the current temperature value against the threshold. If it above the high threshold or under the low threshold, alarm shall be raised.
+
+Besides device temperature, shall also monitoring the fan running status.
+
+Thermal device monitoring will loop at a certain period, 60s can be a good value since usually temperature don't change much in a short period.
+
+### 2.1 Temperature monitoring 
+
+In new platform API ThermalBase() class provides get_temperature(), get_high_threshold(),  get_low_threshold(), get_critical_high_threshold() and get_critical_low_threshold() functions, values for a thermal object can be fetched from them. Warning status can also be deduced. 
+
+For the purpose of feeding CLI/SNMP or telemetry functions, these values and warning status can be stored in the state.  DB schema can be like this:
+
+    ; Defines information for a thermal object
+    key                     = TEMPERATURE_INFO|object_name   ; name of the thermal object(CPU, ASIC, optical modules...)
+    ; field                 = value
+    temperature             = FLOAT                          ; current temperature value
+    timestamp               = STRING                         ; timestamp for the temperature fetched
+    high_threshold          = FLOAT                          ; temperature high threshold
+    critical_high_threshold = FLOAT                          ; temperature critical high threshold
+    low_threshold           = FLOAT                          ; temperature low threshold
+    critical_low_threshold  = FLOAT                          ; temperature critical low threshold
+    warning_status          = BOOLEAN                        ; temperature warning status
+
+These devices shall be included to the temperature monitor list but not limited to: CPU core, CPU pack, ASIC, PSU, Optical Modules, etc.
+
+TEMPERATURE_INFO Table key object_name convention can be "device_name + index"  or device_name if there is no index, like "cpu_core_0", "asic", "psu_2".  Appendix 1 listed all the thermal sensors that supported on Mellanox platform.
+    
+### 2.2 Fan device monitoring 
```

Comment:
It would be desirable to monitor both the FAN tray's FANs as well as the PSU's FANs.

Reply:
both of them will be monitored.

Reply:
ok thanks